airbyte-integrations/connectors/destination-mongodb/src/main/java/io/airbyte/integrations/destination/mongodb/MongodbRecordConsumer.java

```diff
@@ -0,0 +1,183 @@
+      final AirbyteRecordMessage recordMessage = message.getRecord();
+      AirbyteStreamNameNamespacePair pair = AirbyteStreamNameNamespacePair.fromRecordMessage(recordMessage);
+
+      if (!writeConfigs.containsKey(pair)) {
+        LOGGER.error("Message contained record from a stream that was not in the catalog. catalog: {}, message: {}",
+            Jsons.serialize(catalog), Jsons.serialize(recordMessage));
+        throw new IllegalArgumentException(
+            String.format("Message contained record from a stream that was not in the catalog. \ncatalog: %s , \nmessage: %s",
+                Jsons.serialize(catalog), Jsons.serialize(recordMessage)));
+      }
+      var writeConfig = writeConfigs.get(pair);
+      insertRecordToTmpCollection(writeConfig, message, recordMessage);
+    }
+  }
+
+  @Override
+  protected void close(boolean hasFailed) {
+    try {
+      if (!hasFailed) {
+        LOGGER.info("Migration finished with no explicit errors. Copying data from tmp tables to permanent");
+        writeConfigs.values().forEach(mongodbWriteConfig -> Exceptions.toRuntime(() -> {
+          try {
+            copyTable(mongoDatabase, mongodbWriteConfig.getCollectionName(), mongodbWriteConfig.getTmpCollectionName());
+          } catch (RuntimeException e) {
+            LOGGER.error("Failed to process a message for Streams numbers: {}, \nSyncMode: {}, \nCollectionName: {}, \nTmpCollectionName: {}",
+                catalog.getStreams().size(), mongodbWriteConfig.getSyncMode(), mongodbWriteConfig.getCollectionName(),
+                mongodbWriteConfig.getTmpCollectionName());
+            LOGGER.error("Failed with exception: {}", e.getMessage());
+            throw new RuntimeException(e);
+          }
+        }));
+        outputRecordCollector.accept(lastStateMessage);
+      } else {
+        LOGGER.error("Had errors while migrations");
+      }
+    } finally {
+      LOGGER.info("Removing tmp collections...");
+      writeConfigs.values()
+          .forEach(mongodbWriteConfig -> mongoDatabase.getCollection(mongodbWriteConfig.getTmpCollectionName()).drop());
+      LOGGER.info("Finishing destination process...completed");
+    }
+  }
+
+  /* Helpers */
+
+  private void insertRecordToTmpCollection(MongodbWriteConfig writeConfig,
+                                           AirbyteMessage message,
+                                           AirbyteRecordMessage recordMessage) {
+    try {
+      Map<String, Object> result = objectMapper.convertValue(recordMessage.getData(), new TypeReference<>() {});
+      var newDocument = new Document();
+      newDocument.put(AIRBYTE_DATA, new Document(result));
+      newDocument.put(AIRBYTE_EMITTED_AT, new LocalDateTime().toString());
+
+      var collection = writeConfig.getCollection();
+
+      var documents = writeConfig.getDocuments();
+      if (!documents.contains(newDocument.get(AIRBYTE_DATA, Document.class))) {
+        collection.insertOne(newDocument);
+      }
```

Comment:
I suppose this check is to prevent duplicated data. However, how efficient is this check if the collection is huge?

Reply:
updated it